observability/tracing/provider.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2025 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracing
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+)
+
+func noopFunc(context.Context) error { return nil }
+
+type TracerProvider struct {
+	trace.TracerProvider
+	shutdown func(context.Context) error
+}
+
+func (m *TracerProvider) Shutdown(ctx context.Context) error {
+	return m.shutdown(ctx)
+}
+
+func DefaultTextMapPropagator() propagation.TextMapPropagator {
+	return propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	)
+}
+
+func NewTracerProvider(
+	ctx context.Context,
+	cfg Config,
+	opts ...sdktrace.TracerProviderOption,
+) (*TracerProvider, error) {
+	if cfg.Protocol == ProtocolNone {
+		return &TracerProvider{
+			TracerProvider: noop.NewTracerProvider(),
+			shutdown:       noopFunc,
+		}, nil
+	}
+
+	exp, err := exporterFor(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating tracer exporter: %w", err)
+	}
+
+	sampler, err := sampleFor(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating tracer sampler: %w", err)
+	}
+
+	provider := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exp),
+		sdktrace.WithSampler(sampler),
+	)
+	return &TracerProvider{
+		TracerProvider: provider,
+		shutdown:       provider.Shutdown,
+	}, nil
+}
+
+func exporterFor(ctx context.Context, cfg Config) (sdktrace.SpanExporter, error) {
+	switch cfg.Protocol {
+	case ProtocolGRPC:
+		otlptracegrpc.New(ctx)
```

Comment:
Is there a reason we call this once here and then again in `buildGRPC`?

Reply:
that's a mistake